### PR TITLE
Implement tickets 153-158

### LINKED ISSUE #153: Fix Claude Code workspace wiring when .claude/agents already exists
# Summary

A real-workspace release-confidence pass found that `agent-harness workspace claude-code` can fail during workspace wiring with `EEXIST` while trying to create `.claude/agents`.

This is not a release-harness artifact: discovery, selection, recommendation, mirror, install, and activation all completed, and the failure happens specifically during the Claude Code native wire phase.

## Audit run context

Release audit root:

- `C:\Users\ar271\.openclaw\workspace\release-audits\agent-harness-v1.0.3-1778338295156`

Failing case:

- workspace repo: `C:\Projects\InterActNote`
- host: `claude-code`
- log: `C:\Users\ar271\.openclaw\workspace\release-audits\agent-harness-v1.0.3-1778338295156\interactnote-claude-code\logs\02-workspace.log`

## Reproduction

From a real repo worktree of `C:\Projects\InterActNote`, the audit ran:

```bash
node C:\Projects\agent-harness\dist\cli.js --no-dotenv workspace claude-code
```

The run progressed through:

- demand profiling
- source indexing
- catalog selection
- recommendation generation
- mirror acquisition
- install reconciliation
- activation view generation

Then failed during wire/apply with:

```text
Error: EEXIST: file already exists, mkdir 'C:\Users\ar271\AppData\Local\Temp\agent-harness-v103-audit-IyMNGS\interactnote-claude-code-worktree\.claude\agents'
    at async mkdir (node:internal/fs/promises:859:10)
    at async ensureDirectory (file:///C:/Projects/agent-harness/dist/files.js:83:5)
    at async writeTextFile (file:///C:/Projects/agent-harness/dist/files.js:157:5)
    at async writeClaudeCodeNativeFiles (file:///C:/Projects/agent-harness/dist/host-adapters/native-wire.js:429:5)
    at async wireNativeHost (file:///C:/Projects/agent-harness/dist/host-adapters/native-wire.js:135:34)
    at async runWorkspace (file:///C:/Projects/agent-harness/dist/workspace.js:49:5)
```

## Why this matters

This blocks a real `workspace claude-code` flow on a real repo and looks like a native wire path bug, likely around handling an already-existing path that is not in the expected directory/link state.

## Acceptance criteria

- `agent-harness workspace claude-code` succeeds on the InterActNote-style real repo case used in the audit.
- Creating or reusing `.claude/agents` is idempotent/safe when the path already exists.
- Claude Code native wire/apply no longer throws `EEXIST` for this path.
- Add regression coverage for repeated/apply-on-existing-path behavior in the Claude Code adapter.

### LINKED ISSUE #154: Fix OpenCode/Pi reset leaving AGENTS.md modified after reset
# Summary

A real-workspace release-confidence pass found that `wire --reset` for the OpenCode and Pi adapters can leave `AGENTS.md` modified after reset on a real repo worktree.

The reset commands themselves returned success, but the post-reset git status still showed `M AGENTS.md`.

## Audit run context

Release audit root:

- `C:\Users\ar271\.openclaw\workspace\release-audits\agent-harness-v1.0.3-1778338295156`

Affected cases:

### OpenCode
- workspace repo: `C:\Projects\Apify\Webhook Debugger and Logger`
- host: `opencode`
- reset log: `C:\Users\ar271\.openclaw\workspace\release-audits\agent-harness-v1.0.3-1778338295156\apify-opencode\logs\04-wire-reset.log`
- post-reset status log: `C:\Users\ar271\.openclaw\workspace\release-audits\agent-harness-v1.0.3-1778338295156\apify-opencode\logs\05-git-status.log`

### Pi
- workspace repo: `C:\Projects\Apify\Webhook Debugger and Logger`
- host: `pi`
- reset log: `C:\Users\ar271\.openclaw\workspace\release-audits\agent-harness-v1.0.3-1778338295156\apify-pi\logs\04-wire-reset.log`
- post-reset status log: `C:\Users\ar271\.openclaw\workspace\release-audits\agent-harness-v1.0.3-1778338295156\apify-pi\logs\05-git-status.log`

## Reproduction shape

The audit ran successful reset commands:

```bash
node C:\Projects\agent-harness\dist\cli.js --no-dotenv wire opencode --reset
node C:\Projects\agent-harness\dist\cli.js --no-dotenv wire pi --reset
```

Both commands exited successfully, but `git status --short` afterwards still reported:

```text
 M AGENTS.md
```

## Why this matters

For project-local adapters that manage sections in `AGENTS.md`, reset should restore the workspace to a clean baseline when the harness-managed content is removed. Leaving `AGENTS.md` modified after a nominally successful reset makes the adapter feel non-reversible and is a release-quality problem even when the core operation succeeds.

## Acceptance criteria

- `wire opencode --reset` restores `AGENTS.md` to a clean post-reset state in the audited Apify-style real repo case.
- `wire pi --reset` restores `AGENTS.md` to a clean post-reset state in the same case.
- If the repo already contains unmanaged `AGENTS.md` content, reset preserves that content without leaving a diff from the pre-apply baseline.
- Add regression coverage for apply/reset round-trips on adapters that manage `AGENTS.md` sections.

### LINKED ISSUE #155: Reduce selection and ranking false positives for broad security/platform signals
# Summary

A manual review of the selected catalog for the real `agent-harness` workspace shows that selection is still too permissive for broad devtools/security/platform signals. The selected set includes clearly unrelated vertical/vendor-specific assets such as Firebase, Power Platform, Dataverse, Power BI, and Azure-enterprise assets even though the workspace signals do not contain those exact domains.

This looks like selection-stage leakage, not just ranking noise.

## Evidence

Attached/inspected selected catalog:

- `C:\Users\ar271\.openclaw\media\inbound\catalog.selected---7dcf9d39-7f62-4b58-bba2-205682727e7c.jsonl`

Real workspace demand profile used for comparison:

- `C:\Users\ar271\.openclaw\workspace\release-audits\agent-harness-v1.0.3-1778338295156\agent-harness-cursor\state\discover\output\demand-profile.json`

Actual workspace signals were:

- languages: javascript, typescript
- packageManagers: npm
- frameworks: (none)
- concerns: ci-cd, devops, documentation, infrastructure, integration, knowledge-base, platform-engineering, security
- tooling: detector:devops-platform, detector:documentation, node, npm:@eslint/js, npm:eslint, npm:eslint-config-prettier, npm:globals, npm:husky, npm:jsonc-parser, npm:lint-staged, npm:prettier, npm:typescript, npm:typescript-eslint, typescript

Notably absent:

- Firebase
- Power Platform / Power Apps / Dataverse
- Power BI
- Azure-specific cloud stack
- Kubernetes / Helm / cluster orchestration specifics

## False-positive clusters found in selected catalog

- Firebase: 3
- Power Platform / Dataverse / Power BI: 17
- Azure-specific enterprise stack: 18
- Kubernetes / cluster-specific platform work: 1

### Sample clearly false-positive entries

- Firebase Security Rules Auditor (firebase-skills:skills%2Ffirebase-security-rules-auditor%2FSKILL.md)
- Power Platform Expert.agent (github-awesome-copilot:agents%2Fpower-platform-expert.agent.md)
- Power Platform Mcp Integration Expert.agent (github-awesome-copilot:agents%2Fpower-platform-mcp-integration-expert.agent.md)
- Dataverse Python Authentication Security.instructions (github-awesome-copilot:instructions%2Fdataverse-python-authentication-security.instructions.md)
- Power Bi Security Rls Best Practices.instructions (github-awesome-copilot:instructions%2Fpower-bi-security-rls-best-practices.instructions.md)
- Power Platform Connector.instructions (github-awesome-copilot:instructions%2Fpower-platform-connector.instructions.md)
- Power Platform Mcp Development.instructions (github-awesome-copilot:instructions%2Fpower-platform-mcp-development.instructions.md)
- Power Platform Expert (github-awesome-copilot:plugins%2Fpower-apps-code-apps%2Fagents%2Fpower-platform-expert.md)
- Plugin (github-awesome-copilot:plugins%2Fpower-platform-architect%2F.github%2Fplugin%2Fplugin.json)
- Power Platform Architect (github-awesome-copilot:plugins%2Fpower-platform-architect%2FREADME.md)
- Power Platform Architect (github-awesome-copilot:plugins%2Fpower-platform-architect%2Fskills%2Fpower-platform-architect%2FSKILL.md)
- Plugin (github-awesome-copilot:plugins%2Fpower-platform-mcp-connector-development%2F.github%2Fplugin%2Fplugin.json)

### Sample borderline entries

- Platform Sre Kubernetes.agent (github-awesome-copilot:agents%2Fplatform-sre-kubernetes.agent.md)

## Why this matters

Generic concerns like `security`, `platform-engineering`, `integration`, and `infrastructure` are currently enough to admit highly specific vendor/domain assets that do not match the workspace's actual stack.

This pollutes both:

- manual review of `catalog.selected`
- the recommendation pool that ranking has to clean up later

## Ranking leakage snapshot

- shared: 0 flagged entries in top 20
- copilot-vscode: 5 flagged entries in top 20
- opencode: 5 flagged entries in top 20
- cursor: 5 flagged entries in top 20
- zed: 3 flagged entries in top 20
- claude-code: 5 flagged entries in top 20
- pi: 3 flagged entries in top 20

## Likely causes

1. Selection is admitting vendor/domain-specific assets on broad umbrella concerns alone.
2. Official-source trust/priority is rescuing weak semantic matches.
3. Token overlap on words like `security`, `platform`, `integration`, and `mcp` is too permissive.
4. Support/reference files inside broad vendor repos are surviving as standalone selected assets too easily.

## Acceptance criteria

- Vendor/domain-specific assets require exact or near-exact medium/strong evidence before surviving selection.
- Generic `security` / `platform` / `integration` signals alone do not admit Firebase, Power Platform, Dataverse, Power BI, Azure-enterprise, or similar vertical assets.
- Kubernetes/cluster-specific assets require actual container/orchestration evidence before selection.
- Support/reference files like repo READMEs, plugin manifests, and examples are not selected as first-class assets unless they have strong direct fit.
- Add regression coverage using an `agent-harness`-style workspace fixture so this false-positive pattern stays fixed.

### LINKED ISSUE #156: Prevent installed dependency evidence from self-recommending package assets
# Summary

A real Apify workspace audit shows a dependency-evidence feedback loop: already-installed package dependencies are being reused as strong recommendation evidence, which can cause the recommender to surface package-registry assets back to the user based on the same package that was already present in the workspace.

DuckDB is the clearest example. The workspace contains `npm:@duckdb/node-api`, and the recommendation model then treats `npm-duckdb-node-api` as a strong tooling signal in ranking. This is evidence of stack, but not evidence that the user needs the same package suggested back to them.

# Evidence

Workspace demand profiles:
- `C:\Users\ar271\.openclaw\media\inbound\demand-profile---3845a48d-2c8a-435f-a5e8-f95081ebd639.json`
- `C:\Users\ar271\.openclaw\media\inbound\demand-profile---bc586512-5628-41d9-8812-a5ec12e5a0a4.json`

Recommendation reports:
- `C:\Users\ar271\.openclaw\media\inbound\recommendations---f817a512-ce64-4b0f-94d0-45c13029a5d1.json`
- `C:\Users\ar271\.openclaw\media\inbound\recommendations---c4dba6ea-e4e2-477f-91a6-fecf868c77ee.json`

Workspace has DuckDB dependency signal: true

Examples where the ranking uses the DuckDB dependency signal:
- shared #1: npm-registry:npm:%40apify%2Factors-mcp-server
- shared #2: npm-registry:npm:%40iflow-mcp%2Fapify-apify-mcp-server
- shared #3: npm-registry:npm:hono-mcp-server-sse-transport
- shared #4: npm-registry:npm:%40iflow-mcp%2Fapify-mcp-server
- shared #5: npm-registry:npm:%40mindstone-engineering%2Fmcp-server-elevenlabs
- shared #6: npm-registry:npm:%40dominusnode%2Fmcp-server
- copilot-vscode #1: npm-registry:npm:%40apify%2Factor-memory-expression
- copilot-vscode #2: npm-registry:npm:%40duckdb%2Fnode-api

# Why this matters

Dependency presence should help infer stack/domain (for example, that the repo uses DuckDB or data tooling), but it should not directly cause package-registry or package-like assets for that same dependency to be suggested back to the user unless they provide clearly distinct new capability.

This makes the recommender feel tautological: the package was detected because it was already installed, and then it is suggested because it was detected.

# Acceptance criteria

- Installed dependency evidence does not directly self-recommend the same package or package-equivalent asset back to the user.
- Dependency evidence can still contribute to domain inference (for example DuckDB -> data / analytics), without causing package self-echo.
- Distinct capability assets in the same domain (skills, MCP servers, instructions, workflows) may still rank when they add new value.
- Add regression coverage for a workspace that already contains a package dependency and verify the recommender does not simply suggest that dependency back.

### LINKED ISSUE #157: Improve design-tool recall for design-system-heavy workspaces
# Summary

A real InterActNote workspace audit suggests a recall gap for design-system/design-asset tooling: the workspace emits strong design-oriented concerns, but Penpot-related assets do not appear in selected catalog or final ranking.

At the same time, broad mobile signals dominate the recommendations, so the system may be over-indexing on generic mobile MCP/package assets while under-surfacing likely design-collaboration tools.

# Evidence

Workspace demand profile:
- `C:\Users\ar271\.openclaw\media\inbound\demand-profile---f6b99995-7fbe-4c9f-b7dc-46a13907c488.json`

Recommendation report:
- `C:\Users\ar271\.openclaw\media\inbound\recommendations---b8a3b547-3b56-4a1f-9722-55ad3b069017.json`

Key workspace signals:
- frameworks: flutter
- concerns include: design-assets, design-systems, frontend, mobile
- tooling includes: detector:design-system, detector:mobile, flutter, pub

Penpot presence:
- penpot mentioned in demand profile: false
- penpot in selected catalog: 0
- penpot in final ranking: 0

# Why this matters

For a still-green workspace, the recommender should be especially careful not to collapse all design/mobile evidence into generic mobile stacks only. If design-system/design-assets evidence is present, relevant design tooling domains (for example Penpot) should at least be discoverable candidates when such assets exist in the source universe.

# Acceptance criteria

- Review whether Penpot-related assets are present in the source universe and, if so, why they were not selected or ranked for this workspace.
- Design-system / design-assets signals can contribute to recall for relevant design tooling domains rather than only generic mobile assets.
- Add regression coverage for a green Flutter/design-oriented workspace where design-system evidence is present.

### LINKED ISSUE #158: Add AI-assisted review and reranking for workspace-aware recommendations
# Summary

Add a bounded AI-assisted review/rerank stage for workspace-aware recommendations.

The current deterministic discovery and recommendation pipeline already does the hard infrastructure work well: signal extraction, source discovery, mirroring, compatibility, trust, quarantine, install, activation, and wiring. The weak point is the judgment layer between selected catalog and final ranking.

The existing `discover enrich` feature is not sufficient for that role. Today it is an optional OpenAI-compatible summary sidecar that reads the demand profile + selected catalog and writes `discover/output/ai-enrichment.json`; it does not alter ranking, suppress false positives, recover missing domains, or feed back into recommendation decisions.

## Proposal

Introduce a recommendation-native AI review stage, for example:

```bash
agent-harness recommend ai-review
agent-harness recommend report --ai-review
```

Design principles:
- keep deterministic discovery as the candidate generator
- use AI only on a bounded shortlist
- treat AI as a reviewer/reranker, not as an unconstrained asset discoverer
- preserve deterministic post-AI validation and policy enforcement

### Inputs to AI review
- workspace demand summary
  - top signals
  - evidence-strength histograms
  - scan truncation indicators
  - strongest evidence paths
- deterministic candidate shortlist
  - asset id, kind, source family, authority tier, host support
  - matched signals
  - reasons
  - score breakdown
  - duplicate group
  - local availability
  - trust/risk markers
  - optional installed-dependency overlap flags
- bounded mirror snippets where available
- target host / intent

### AI review responsibilities
Allowed:
- keep / questionable / suppress decisions on shortlisted assets
- bounded rerank adjustments
- duplicate spam detection
- self-echo suspicion flags
- missing-domain suggestions based on existing candidate universe

Disallowed:
- inventing assets outside the deterministic candidate universe
- bypassing compatibility / quarantine / trust policy
- mutating install / wire state directly

### Output artifact
Persist a structured artifact such as:
- `recommend/output/ai-review-input.json`
- `recommend/output/ai-review.json`

The output should include:
- reviewed hosts
- accepted asset ids
- questionable assets with rationale + confidence
- suppressed asset ids
- rerank deltas
- missing domains / warnings

### Post-AI enforcement
After the AI response:
- validate schema
- ignore unknown asset ids
- clamp rerank deltas / suppressions to policy limits
- preserve host caps, duplicate caps, budget caps, and safety rules
- then emit the final recommendation report

## Why this matters

This direction directly targets the failure modes already found in real audits:
- `#155` broad false-positive selection/ranking leakage
- `#156` installed-dependency self-echo loops
- `#157` missing design-tool recall for design-system-heavy workspaces

It also gives `agent-harness` a stronger product shape:
- deterministic workspace intelligence + policy + safety backbone
- AI-assisted recommendation judgment
- deterministic enforcement before anything becomes actionable

## Acceptance criteria

- Add a recommendation-native AI review command path (`recommend ai-review` and/or `recommend report --ai-review`).
- Reuse the configured AI endpoint machinery, but move the feature from `discover enrich` summary mode into recommendation review/rerank mode.
- AI review operates only on a bounded deterministic shortlist.
- AI review writes structured auditable artifacts.
- Final recommendation output remains policy-validated and host-safe.
- Add evaluation coverage using the existing real-workspace failure patterns from `#155`, `#156`, and `#157`.

### PR DESCRIPTION
**Pull request by ar27111994**
## Summary

This PR completes the remaining open tickets #153 through #158.

### Native wiring and reset correctness
- fix native directory handling so existing usable directory links/junctions do not break wiring
- add exact text-file snapshots to wire plans so reset/rollback restores original managed markdown/text files byte-for-byte
- validate managed snapshot paths during reset so hostile or malformed wire plans cannot restore outside the managed file set
- improve OpenCode / Claude Code / Pi native cleanup behavior
- harden Windows build cleanup against transient `dist/` removal contention

### Recommendation precision and recall
- suppress specialized false positives from broad generic demand signals
- improve design-tool recall for design-system / design-asset demand
- prevent dependency self-echo where installed package evidence recommends the same package asset back to the workspace
- add richer evaluation fixtures and targeted regression tests for these edge cases

### AI-assisted recommendation review
- add bounded recommendation-native AI review flow
- support:
  - `recommend ai-review`
  - `recommend report --ai-review`
- write validated AI review artifacts under `recommend/output/`
- sanitize and bound AI review response handling before applying reranks/suppressions

### Docs and examples
- update README command examples and behavior notes
- document AI review environment/config usage in `.env.example`
- update CLI help text

## Validation

- `npm audit --omit=dev` → 0 vulnerabilities
- `npm run validate:quality`
- `npm run validate:recommendations` → 12/12 fixtures passed

## Summary by CodeRabbit

* **New Features**
  * Added bounded AI review workflow for recommendations, accessible via new `recommend ai-review` command with `--apply` option.
  * Enhanced catalog entry filtering with specialized demand-relevance gates for better precision.
  * Implemented text file snapshot and restoration capability for improved failure recovery.

* **Documentation**
  * Updated README with AI review examples and clarified shared configuration between enrichment and review workflows.

* **Bug Fixes**
  * Improved directory creation error handling and managed section content preservation.

* **Chores**
  * Enhanced build script reliability with better retry mechanisms.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ar27111994/agent-harness/pull/159)

### CODE REVIEW

File: src/recommend/commands.ts [hunk 1]
```diff
@@ -29,14 +30,35 @@ export async function runRecommend(
 
   switch (command) {
     case "report": {
-      const report = await writeRecommendationReport(projectRoot);
+      const policy = await loadRecommendationPolicy(projectRoot);
+      const deterministicReport = await writeRecommendationReport(projectRoot);
+      const report = rest.includes("--ai-review")
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/recommend/commands.ts]: Fixed in fd2ecbe. `recommend report` now only loads the policy when `--ai-review` is requested, and both `report` and `ai-review` reuse the preloaded policy when writing the deterministic report.

File: src/tests/files.test.ts [hunk 2]
```diff
@@ -1,109 +1,59 @@
...
 
-void test("recursive scan stops at explicit file budgets", async () => {
-  const root = await mkdtemp(join(tmpdir(), "agent-harness-scan-"));
-  try {
-    await writeFile(join(root, "one.txt"), "one", "utf8");
-    await writeFile(join(root, "two.txt"), "two", "utf8");
+void test("ensureDirectory accepts an existing managed directory link", async () => {
+  const root = await mkdtemp(join(tmpdir(), "agent-harness-files-test-"));
+  const targetDirectory = join(root, "target");
+  const linkedDirectory = join(root, "linked");
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/files.test.ts]: Fixed in fd2ecbe. Added `try/finally` cleanup with `rm(..., { recursive: true, force: true })` for the temp root.

File: src/recommend/signals.ts [hunk 3]
```diff
@@ -134,6 +140,83 @@ export function shouldEnforceConcernTarget(
...
+    keywords.add("penpot");
+  }
+
+  return keywords;
+}
+
+function normalizePackageManifestEntry(value: string): string | null {
+  const match = value.match(
+    /^(?:cargo|gem|go|maven|npm|nuget|packagist|pub|pypi|swift):(.+)$/u,
+  );
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/recommend/signals.ts]: Fixed in fd2ecbe. Package-manifest prefix parsing is now centralized in `src/lib/package-manifest-entry.ts` and reused from both discovery selection and recommendation signals so the supported prefixes stay in sync.

File: src/recommend/candidates.ts [hunk 4]
```diff
@@ -85,6 +120,17 @@ export function buildCandidateRecommendation(
...
+    isSuppressedBySpecializedDemandGate(entry, rawKeywordTerms, demandContext)
+  ) {
+    return null;
+  }
+  if (isSuppressedForDesignSystemDemand(rawKeywordTerms, demandContext)) {
+    return null;
+  }
+  if (isSuppressedByDependencySelfEcho(entry, demandContext)) {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/recommend/candidates.ts:105]: Fixed in fd2ecbe. Removed the unreachable dependency self-echo penalty/reason path so suppression is the single mechanism.

File: src/recommend/candidates.ts [hunk 5]
```diff
@@ -30,6 +30,34 @@ const WRAPPER_LIKE_TERMS = new Set([
...
+      ["power", "bi"],
+    ],
+    entryTerms: [
+      ["dataverse"],
+      ["power", "platform"],
+      ["power", "apps"],
+      ["power", "bi"],
+    ],
+  },
+] as const;
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/recommend/candidates.ts]: Fixed in fd2ecbe. The specialized-demand gates now live in `src/domains/discovery/demand-helpers.ts` and are shared by both deterministic selection and recommendation suppression.

File: src/host-adapters/native-wire.ts [hunk 6]
```diff
@@ -1006,6 +1038,91 @@ async function cleanupFailedNativeHostApply(
   }
 }
 
+function resolveManagedTextFileSnapshotPaths(
+  spec: NativeHostSpec,
+  workspaceRoot: string,
+): string[] {
+  switch (spec.host) {
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/host-adapters/native-wire.ts:1076]: Fixed in fd2ecbe. Native text snapshot coverage now includes `.rules`, `CLAUDE.md`, and `.claude/CLAUDE.md`, and reset/failure cleanup restores those exact snapshots byte-for-byte.

File: src/recommend/ai-review.ts [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+
+  const firstChoice = asJsonObject(choicesValue[0]);
+  const message = asJsonObject(firstChoice?.message);
+  const content = extractAiReviewMessageContent(message);
+
+  try {
+    return JSON.parse(content) as unknown;
+  } catch {
+    return {};
+  }
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/recommend/ai-review.ts:363]: Fixed in fd2ecbe. Malformed model JSON now throws inside the existing `try` path, so the run writes a failed artifact instead of a misleading completed one.

File: src/tests/wire-plan-security.test.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+import { readTextFileOrNull, writeJsonFile, writeTextFile } from "../files.js";
+import { wireNativeHost } from "../host-adapters/native-wire.js";
+import { wireOpenCode } from "../host-adapters/opencode.js";
+
+void test("native reset rejects wire-plan text snapshots outside the managed restore set", async () => {
+  const root = await mkdtemp(
+    join(tmpdir(), "agent-harness-native-wire-security-"),
+  );
+  const projectRoot = join(root, "project");
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/wire-plan-security.test.ts]: Fixed in fd2ecbe. Added `try/finally` cleanup for the temp root used by this security test.

File: src/tests/wire-plan-security.test.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    /outside the managed restore set/u,
+  );
+  assert.equal(await readTextFileOrNull(escapedPath), "do-not-touch");
+});
+
+void test("OpenCode reset rejects wire-plan text snapshots outside the managed restore set", async () => {
+  const root = await mkdtemp(
+    join(tmpdir(), "agent-harness-opencode-security-"),
+  );
+  const projectRoot = join(root, "project");
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/tests/wire-plan-security.test.ts]: Fixed in fd2ecbe. Added `try/finally` cleanup for the temp root used by this second security test as well.

File: src/manifest-validation/workspace.ts [hunk 10]
```diff
@@ -139,9 +139,29 @@ export function assertWirePlanManifest(
...
+  }
+
+  value.forEach((entry, index) => {
+    const snapshot = assertRecord(entry, `${context}[${index}]`);
+    assertString(snapshot.path, `${context}[${index}].path`);
+    if (snapshot.content !== null) {
+      assertString(snapshot.content, `${context}[${index}].content`);
+    }
+  });
+}
```
[the remark this answers is not in the document]
Comment (ar27111994) [reply src/manifest-validation/workspace.ts:169]: Fixed in fd2ecbe. `assertTextFileSnapshots` now rejects duplicate snapshot paths while keeping the existing content null/string validation unchanged.
